### ISSUE TITLE
Support mocking the return value when the type is Task/Task<T>

### ISSUE THREAD

**Issue by JasonMing** [moq/moq4#171]
When `Mock<T>.DefaultValue == DefaultValue.Empty`, the default return value of async method, which returns `Task` or `Task<T>`, will directly return null, then `await` on them will cause NullRefereceException.
In this case, the "Empty" value should be `Task.FromResult(default(T))`.

When `Mock<T>.DefaultValue == DefaultValue.Mock` the default return value of async method will change to the mocking `Task`/`Task<T>`, however, the `Task`/`Task<T>` are not able to be mocked, then the **System.ArgumentExceptionType to mock must be an interface or an abstract or non-sealed class.** will be thrown.
In this case, the "Mock" value should be `Task.FromResult(Mock.Of<T>())`.

**Comment by stakx** [comment]
> When `Mock<T>.DefaultValue == DefaultValue.Empty` [...], the "Empty" value should be `Task.FromResult(default(T))`.

This is already implemented.

> When `Mock<T>.DefaultValue == DefaultValue.Mock` [...], the "Mock" value should be `Task.FromResult(Mock.Of<T>())`.

Agreed. I'm looking at this now.

> however, the `Task`/`Task<T>` are not able to be mocked, then the System.ArgumentExceptionType to mock must be an interface or an abstract or non-sealed class. will be thrown.

Somehow this sentence became a little garbled. I suppose you meant, "The type to mock must be an interface or an abstract or non-sealed class. If it is not able to be mocked, a System.ArgumentException will be thrown."?

The new logic is likely going to be this when `DefaultValue == DefaultValue.Mock` and `Task<T>` is to be mocked:

1. Let the empty default value provider produce a default value for `T`.
2. If that default value is not `null`, return it.
3. Otherwise, if `T` is a mockable type, mock it and return a completed task containing the generated mock.
4. Otherwise, let the empty default value provider generate a completed task containing `default(T)`.

**Comment by stakx** [comment]
This turned out to be quite difficult to implement, and this feature comes with a small runtime performance penalty, so some more optimization might be required later on.

**Comment by stakx** [comment]
@JasonMing, @westonal, @kibiz0r, @fadookie, @jahmai, @almazik- I just published a [pre-release version 4.8.0-rc1 on NuGet](https://www.nuget.org/packages/Moq/4.8.0-rc1). It has (among other things) support for `Task<>` and `ValueTask<>` with both `DefaultValue.Mock` and `DefaultValue.Empty`. Feel free to test away.

**Comment by stakx** [comment]
@jahmai - You're not hallucinating, but `Task` support is more complete now. Moq has been happily wrapping default empty values (`DefaultValue.Empty`) in a completed task for a while. The same behavior now works with `DefaultValue.Mock`, too:

```csharp
public abstract class Inner { }

public abstract class Outer
{
    public abstract Task<Inner> GetInnerAsync();
}

var mock = new Mock<Outer> { DefaultValue = DefaultValue.Mock };
var inner = await mock.Object.GetInnerAsync();
Assert.NotNull(inner); // NEW: this assertion used to fail, but now it passes
```

(Same for `ValueTask<>`.)
